### PR TITLE
changing xDS APIs to take abs::Satus

### PR DESCRIPTION
**Pull request by alyssawilk**
Unlike many of the throw->status PRs I'm doing xDS in two parts - one is a pure API change, and the follow-up(s) will change the throw over to returning non-Ok stauts.  This PR is large enough already I didn't want to do them all at once.

Risk Level: low (but will cause downstream churn, sorry)
Testing: updated tests
Docs Changes: n/a
Release Notes: n/a
https://github.com/envoyproxy/envoy-mobile/issues/176

### CODE REVIEW

File: source/common/router/vhds.cc [hunk 1]
```diff
@@ -91,11 +91,16 @@ void VhdsSubscription::onConfigUpdate(
               config_update_info_->protobufConfigurationCast().name(),
               config_update_info_->configHash());
     if (route_config_provider_ != nullptr) {
-      route_config_provider_->onConfigUpdate();
+      auto status = route_config_provider_->onConfigUpdate();
+
+      if (!status.ok()) {
+        throw EnvoyException(std::string(status.message()));
+      }
```
Comment (adisuissa) [inline source/common/router/vhds.cc:98]: nit: can THROW_IF_NOT_OK be used here?

File: source/common/filter/config_discovery_impl.cc [hunk 2]
```diff
@@ -215,7 +225,7 @@ void FilterConfigProviderManagerImplBase::applyLastOrDefaultConfig(
   bool last_config_valid = false;
   if (subscription->lastConfig()) {
     TRY_ASSERT_MAIN_THREAD {
-      provider.validateTypeUrl(subscription->lastTypeUrl());
+      THROW_IF_NOT_OK(provider.validateTypeUrl(subscription->lastTypeUrl()))
```
[the remark this answers is not in the document]
Comment (adisuissa) [reply source/common/filter/config_discovery_impl.cc:228]: Looking further in this PR, I guess this is going to be a future change.

File: envoy/common/exception.h [hunk 3]
```diff
@@ -12,6 +12,12 @@ class EnvoyException : public std::runtime_error {
   EnvoyException(const std::string& message) : std::runtime_error(message) {}
 };
 
+#define THROW_IF_NOT_OK(status_fn)                                                                 \
+  absl::Status status = status_fn;                                                                 \
+  if (!status.ok()) {                                                                              \
+    throw EnvoyException(std::string(status.message()));                                           \
+  }
```
Comment (adisuissa) [inline envoy/common/exception.h:19]: Consider adding this inside an inner scope, to allow reuse of `status` variable name in the same block.

File: test/common/filter/config_discovery_impl_test.cc [hunk 4]
```diff
@@ -610,14 +613,16 @@ TYPED_TEST(FilterConfigDiscoveryImplTestParameter, TerminalFilterInvalid) {
...
+    ASSERT_TRUE(config_discovery_test.callbacks_
+                    ->onConfigUpdate(decoded_resources.refvec_, response.version_info())
+                    .ok());
     return;
   }
 
   EXPECT_THROW_WITH_MESSAGE(
-      config_discovery_test.callbacks_->onConfigUpdate(decoded_resources.refvec_,
-                                                       response.version_info()),
+      EXPECT_TRUE(config_discovery_test.callbacks_
```
Comment (adisuissa) [inline test/common/filter/config_discovery_impl_test.cc:623]: A bit confusing (expect throw, while expecting a successful return value), but I guess this is just to satisfy the linter.

Comment (alyssawilk) [reply test/common/filter/config_discovery_impl_test.cc:623]: yeah and they're mostly all changing in the follow-up PR.

**Review by mattklein123 — APPROVED** [review]
LGTM at a high level. I live now in the land of `?` so looking at all of this is pretty painful. 😉

File: source/common/router/scoped_rds.cc [hunk 5]
```diff
@@ -415,26 +415,33 @@ void ScopedRdsConfigSubscription::onConfigUpdate(
...
-      !to_be_removed_rds_providers.empty();
-  ConfigSubscriptionCommonBase::onConfigUpdate();
-  if (any_applied) {
+  auto status_or_applied = addOrUpdateScopes(
+      added_resources, (srds_init_mgr == nullptr ? localInitManager() : *srds_init_mgr),
+      version_info);
+  if (!status_or_applied.status().ok()) {
+    return status_or_applied.status();
+  }
+  bool any_applied = status_or_applied.value();
```
Comment (adisuissa) [inline source/common/router/scoped_rds.cc:433]: nit: const here and places above and below (defer for future PR)

File: envoy/common/exception.h [hunk 6]
```diff
@@ -12,6 +12,14 @@ class EnvoyException : public std::runtime_error {
   EnvoyException(const std::string& message) : std::runtime_error(message) {}
 };
 
+#define THROW_IF_NOT_OK(status_fn)                                                                 \
+  {                                                                                                \
+    absl::Status status = status_fn;                                                               \
```
Comment (adisuissa) [inline envoy/common/exception.h:17]: nit: const (defer to future PR)

**Review by adisuissa — APPROVED** [review]
Thanks, LGTM!
Left a couple of small const nits, that I think can be done in a future PR.

**Comment by alyssawilk** [comment]
Thanks for the fast turnaround! I'll address the last two nits in the follow-up (removing throws) which I have staged to go

**Pull request merged**